### ISSUE TITLE
Investigate and fix possible issues with mixed-precision-training leading to nans

### ISSUE THREAD

**Issue by tanzir5** [odissei-lifecourse/life-sequencing-dutch#142]
Mixed-precision-training (MPT) can lead to nans and gradients becoming 0 even when they are not, leading to early termination of training before convergence. 

More details here: 
https://forum.faceswap.dev/viewtopic.php?t=2177

The two doable possible fixes are:

i) Train without mixed-precision training, let everything be 32 bits.
ii) Use large enough epsilon value for adam optimizer. Details in: https://discuss.pytorch.org/t/adam-half-precision-nans/1765
   

For option i): this is where code needs to be changed: https://github.com/odissei-lifecourse/life-sequencing-dutch/blob/9a0e6ff41f06ed7acccae33e293842491384d64f/pop2vec/llm/src/new_code/pretrain.py#L120

For option ii): change in config files. like here: https://github.com/odissei-lifecourse/life-sequencing-dutch/blob/9a0e6ff41f06ed7acccae33e293842491384d64f/pop2vec/llm/src/new_code/regular_hparams_medium.txt#L39

If training without MPT doesn't make the model converge better, we should further look into the nan issue.

**Comment by tanzir5** [comment]
There are two ways to do mixed precision training. 

i) Specify the precision parameter in the trainer of pytorch-lightning:
https://lightning.ai/docs/pytorch/stable/common/trainer.html#precision
Also see this if interested: https://lightning.ai/docs/pytorch/1.5.9/advanced/mixed_precision.html

We have been using trainer(precision='16-mixed') till now. 

This means all intermediate floating point calculations are done in 16 bits but the weights and gradients are stored as 32 bits. 

ii) torch.set_float32_matmul_precision('medium') see below:

 https://pytorch.org/docs/stable/generated/torch.set_float32_matmul_precision.html 

This only does the matrix multiplications using 16 bits but everything else is done using 32 bits as well as everything is stored in 32 bits. 

In our code, we have been doing both.  Doing (i) essentially means (ii) has zero effect as (ii) is part of (i). 

(ii) is fairly stable and I have not seen people on the internet complaining about nan problems related to it. (i) has led to nan problems in multiple cases, it is not only pytorch-lightning's fault. The common solution for nan problems related to (i) is use large enough epsilon value for the optimizer. Common epsilon value for optimizers are 1e-7, 1e-6 (our case). Float16 cannot is unstable at precision level of 1e-7. So increasing epsilon to 1e-4 is the solution people propose which have worked for others.


I trained five different models and got the following results. 

<html>
<body>
<google-sheets-html-origin><style type="text/css"></style>
mpt | torch.setmatmul | eps | batch/s [Speed]
-- | -- | -- | --
0 | 0 | 1.00E-06 | 1.36
0 | 1 | 1.00E-06 | 2.31
1 | 0 | 1.00E-06 | 2.93
1 | 1 | 1.00E-06 | 2.92
1 | 1 | 1.00E-04 | 2.92


</body>
</html>


The 5th row model still faces nan in epoch 28. It also does not train well beyond epoch 14 (loss decrease rate becomes low). The 2nd row model is the most optimal option for us. It trains very well till the end. Even at epoch 30, the training and validation loss is going slightly down and it terminated because max_epoch=30 was passed. It is 1.3 times slower, but that's what we should be using from now on.

**Comment by tanzir5** [comment]
The csv logs of the pre-training runs can be found on the RA. H\Tanzir_Stuff\training-metrics-precision-test
